[Importer] Add TensorFlowLite importer.

mciprian13 opened the pull request:
**Summary**
- Add importer for TensorFlowLite models. The importer currently supports loading **57** operators out of the **125** operators defined by the TensorFlowLite flatbuffer format.
- Add extra operators for Interpreter and CPU:
  - Sqrt, Rsqrt, Reciprocical, Sin, Cos, CmpNEQ, CmpGT, CmpGTE, And, Or, Xor, Not, etc
- Refactor ArgMax to work with any number of dimensions and add ArgMin implementation.
- Python script to generate TensorFlowLite models for unit testing.

**Documentation**
Mention the TFLite support in AOT.md.

**Test Plan**
Unit tests for the TFLite loader and for the newly added operators.

mciprian13 commented:
@jfix71 This is a bigger PR but before going to review it I need for you to help me decide something. So the TensorFlowLite importer source code has 4 external dependencies which I explicitly put in the path **glow\include\glow\Importer\tflite_flatbuffer**:
- 3 header files from the  [flatbuffer ](https://github.com/google/flatbuffers) repository:
  - **glow\Importer\tflite_flatbuffer\flatbuffers\base.h**
  - **glow\Importer\tflite_flatbuffer\flatbuffers\flatbuffers.h**
  - **glow\Importer\tflite_flatbuffer\flatbuffers\stl_emulation.h**
- 1 header file from the [tensorflow](https://github.com/tensorflow/tensorflow) repository:
  - **glow\Importer\tflite_flatbuffer\schema_generated.h**

They amount to a total of ~23.000 lines of code so the actual number of lines for this PR is much less than the ~29.000 lines of code (actually around 5.000 lines). The questions I have:
-  Whether do you think I should put them into another directory? Also I think the current clang-format build fails because those files are not properly formatted but I don't want to change them and keep them with the original content.
- I was also thinking in making the two repositories as git submodules but they are kind of big and kind of does not make sense for only 4 files ... WDYT?

mciprian13 commented:
@jfix71 I put the files in the `thirdparty` folder as suggested with a `Readme.md`. I think this is ready for review now 😄

jackm321 commented on utils/scripts/gen_tflite_models.py:
```diff
@@ -0,0 +1,958 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
```
Could we include directions for installing any dependencies to run this and how to run it

mciprian13 replied:
I added a comment to list the dependencies and how to run it.

mciprian13 commented on lib/Backends/Interpreter/Interpreter.cpp:
```diff
@@ -182,31 +184,42 @@ bool Interpreter::isOpSupported(const NodeInfo &NI) const {
                 ElemKind::Int32ITy, ElemKind::Int64ITy},
-               {}, {CmpLTNode::ResultIdx}) &&
-           (NI.getOutElemTy(CmpLTNode::ResultIdx) == ElemKind::BoolTy);
+               {}, {CmpEQNode::ResultIdx}) &&
```
Well I don't think it is a problem because all the nodes are similar and they have only one output (ResultIdx = 0). It would make no sense for any of them to diverge from the rest.

mciprian13 commented on lib/Backends/Interpreter/Interpreter.cpp:
```diff
@@ -145,10 +145,12 @@ bool Interpreter::isOpSupported(const NodeInfo &NI) const {
-           (NI.getOutElemTy(ArgMaxNode::ArgmaxIdx) == ElemKind::Int64ITy);
+               {ArgMaxNode::ResultIdx}) &&
+           (NI.getOutElemTy(ArgMaxNode::ResultIdx) == ElemKind::Int64ITy ||
+            NI.getOutElemTy(ArgMaxNode::ResultIdx) == ElemKind::Int32ITy);
```
Well I don't think it is a problem because all the nodes are similar and they have only one output (ResultIdx = 0). It would make no sense for any of them to diverge from the rest.

jackm321 replied:
Ok I guess that seems fine

mciprian13 commented on tools/ClassGen/NodeGen.cpp:
```diff
@@ -170,10 +170,18 @@ int main(int argc, char **argv) {
+      .addMember(MemberType::Unsigned, "Axis")
+      .addMember(MemberType::Boolean, "KeepDims")
+      .addResultFromCtorArg()
+      .setDocstring("Finds index of a minimum element along Axis."
```
Done.

mciprian13 commented on tests/unittests/TFLiteImporterTest.cpp:
```diff
@@ -0,0 +1,227 @@
+using namespace glow;
+
+#include <fstream>
+using namespace std;
```
Done.

mciprian13 commented on tests/unittests/TFLiteImporterTest.cpp:
```diff
@@ -0,0 +1,227 @@
+
+using namespace glow;
+
+#include <fstream>
```
Actually removed since it had no use.

mciprian13 commented on tools/ClassGen/NodeGen.cpp:
```diff
@@ -415,6 +444,100 @@ int main(int argc, char **argv) {
+      .dataParallel()
+      .setDocstring("Performs an element-wise CEIL(x) of the Input operand.");
+
+  BB.newNode("Round")
```
In principle no because all the frameworks (ONNX, Numpy, TensorFlow) use the Round with same rounding mode (round to even meaning that values with fractional part 0.5 are rounded to nearest even integer):
https://www.tensorflow.org/api_docs/python/tf/math/round
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Round

But realizing this I actually remembered that my unit tests did not properly verified the rounding. I updated the unit tests and also replaced the **std::round** with **std::nearbyintf** which uses the correct rounding.

jfix71 commented on tools/ClassGen/NodeGen.cpp:
```diff
@@ -381,32 +389,53 @@ int main(int argc, char **argv) {
+      .setDocstring("Performs an element-wise LESS THAN OR EQUAL comparison "
+                    "between the LHS and RHS operands.");
+
+  BB.newNode("CmpGT")
+      .addInput("LHS")
+      .addInput("RHS")
+      .addResultFromCtorArg()
+      .dataParallel()
+      .setDocstring("Performs an element-wise GREATER THAN comparison between "
+                    "the LHS and RHS operands.");
+
+  BB.newNode("CmpGTE")
+      .addInput("LHS")
+      .addInput("RHS")
+      .addResultFromCtorArg()
+      .dataParallel()
+      .setDocstring("Performs an element-wise GREATER THAN OR EQUAL comparison "
+                    "between the LHS and RHS operands.");
```
Why do we need these? We can always just swap the LHS and RHS and use CmpLT and CmpLTE instead.

mciprian13 replied:
You're right this was kind of dumb. I removed the node/instruction definitions/implementations for `CmpGT` and `CmpGTE` but I left instead in Graph.h/Graph.cpp two convenience functions:
```
  CmpLTNode *createCmpGT(llvm::StringRef name, NodeValue LHS, NodeValue RHS);
  CmpLTENode *createCmpGTE(llvm::StringRef name, NodeValue LHS, NodeValue RHS);
```
to create `CmpLT` and `CmpLTE` nodes with swapped inputs.

jackm321 commented on lib/Backends/CPU/CPUBackend.cpp:
```diff
@@ -323,28 +324,40 @@ bool CPUBackend::isOpSupported(const NodeInfo &NI) const {
-           (NI.getOutElemTy(CmpLTNode::ResultIdx) == ElemKind::BoolTy);
+               {ElemKind::FloatTy, ElemKind::Int8QTy, ElemKind::Int32ITy,
+                ElemKind::Int64ITy},
+               {}, {CmpEQNode::ResultIdx}) &&
```
Oh this was from earlier

mciprian13 commented on lib/Backends/CPU/libjit/libjit.cpp:
```diff
@@ -1477,6 +1511,27 @@ DEFINE_DATA_PARALLEL_KERNEL(libjit_element_pow_kernel_f, float,
+                            std::floor(LHS[idx]))
+DEFINE_DATA_PARALLEL_KERNEL(libjit_element_ceil_kernel_f, float,
+                            std::ceil(LHS[idx]))
+// Rounding mode required by ONNX, Nympy, TensorFlow is round to even which
```
Done here and in other 2 places where I did the same typo.

mciprian13 commented on lib/IR/IRGen.cpp:
```diff
@@ -169,8 +169,8 @@ void IRGenVisitor::post(Node *parent, Node *N) {
-                                      P->getArgmax().getElementType());
-    registerIR(P->getArgmax(), V->getArgmax());
+                                      P->getResult().getElementType());
+    registerIR(P->getResult(), V->getDest());
```
This is how it was before this PR. I changed the instruction of ArgMax to use `.autoIRGen()`.

mciprian13 commented on include/glow/Importer/TFLiteModelLoader.h:
```diff
@@ -0,0 +1,286 @@
+  /// in the original model. This vector contains only the node values (tensors)
+  /// registered in the original model since only those are needed for chaining
+  /// the model graph operators. Other node values created during graph loading
+  /// will node be registered in this vector.
```
Done.

mciprian13 commented on lib/Importer/TFLiteModelLoader.cpp:
```diff
@@ -0,0 +1,2130 @@
+                                getTensorName(tensor).c_str(), dim));
+    shape.push_back(static_cast<dim_t>(dim));
+  }
+  // If tensor shape is empty (scalar) we use a 1D shape with size 1.
+  if (shape.empty()) {
+    shape = {1};
+  }
```
I can't remember the exact problem I had but something did not work properly when using a tensor with empty shape. This works just fine.

mciprian13 commented on lib/Importer/TFLiteModelLoader.cpp:
```diff
@@ -0,0 +1,2130 @@
+void convertUint8ToInt8(const uint8_t *inpPtr, int8_t *outPtr, size_t numElem) {
+  for (size_t idx = 0, idxEnd = numElem; idx < idxEnd; ++idx) {
+    int32_t val = inpPtr[idx];
+    val -= 128;
```
I put a common define in `CommonOperatorLoader.h`:
```
constexpr int32_t UINT8_TO_INT8_SHIFT = 128;
```
and reused this definition in: `TFLiteModelLoader.cpp`, `Caffe2ModelLoader.cpp` and in `PyTorchCommon.h`.

mciprian13 commented:
@jfix71
I agree adding CmpGT and CmpGTE nodes was really dumb (I removed them). While I understand your concern about the operator support for new Backends, I must say that I am not a big fan of the lowering mainly because the lowering always results in performance/memory/precision penalties:
- I just hate to see in a graph a Splat which is converted into a constant with sizes of the order of 128x128x32 (memory)
- Lowering an Abs node to (CmpLE + Select + Mul + Splat) sounds like an overkill. I'm not sure a backend would want to see an Abs implemented like this. (performance)
- Although I am not against the lowering of a floating-point FullyConnected into MatMul + BatchedAdd I am definitely against lowering a quantized FullyConnected since the data precision is reduced to INT8 before adding the bias. (precision)

I would prefer to have these atomic implementations for the CPU without lowering. I come from a world (MCU) where every Byte/microsecond/percent counts.

Currently, the main problem for me is that Glow currently has support for quite a few operators. The ONNX importer I think is less than 50% compliant to the standard. The main reason I introduced the TFLite importer is because the TFLite models are much more constrained (statically typed, no dynamic dimensions) and are always optimized by the TensorFlow framework before conversion (removing unnecessary nodes, constant folding, etc). Hence I expect the TFLite importer to be much more reliable than the ONNX importer.

I think the priority should be adding as much operators as possible with as less lowering as possible for the CPU backend and only later as an addition add lowering logic to make the life easier for other/new Backends.

jfix71 commented:
@mciprian13 

>I would prefer to have these atomic implementations for the CPU without lowering. I come from a world (MCU) where every Byte/microsecond/percent counts.

Yes makes total sense -- perhaps I was unclear, it's totally fine to keep the CPU backend's unlowered implementation. I.e. the ask was just to additionally add lowering logic for some of the more straightforward ones to lower. The CPU backend would return false for them in `shouldLower()`. But the Interpreter wouldn't need to implement them in InterpreterNodes.cpp. This could be done as a followup.

jackm321 commented on lib/Importer/TFLiteModelLoader.cpp:
```diff
@@ -0,0 +1,2110 @@
+    // Check bias scale relative error to inputScale * filterScale.
+    if (biasScale != matMulScale) {
+      float relErr = std::abs(matMulScale - biasScale) / matMulScale;
+      llvm::outs() << opErrMsg(
```
nit stderr here too

mciprian13 replied:
Done.

mciprian13 commented on lib/Importer/TFLiteModelLoader.cpp:
```diff
@@ -0,0 +1,2110 @@
+    // Check bias scale relative error to inputScale * weightsScale.
+    if (biasScale != matMulScale) {
+      float relErr = std::abs(matMulScale - biasScale) / matMulScale;
+      llvm::outs() << strFormat(
```
Done.

mciprian13 commented:
@jackm321 It seems the builds fail on your side ... Is it bad?

jackm321 commented:
@mciprian13 nope not bad, mostly just a couple of build system things, I've fixed it and will land this today. Thanks again for this contribution!

yinghai commented on tests/models/tfliteModels/depthwise_conv2d_c1_m2.inp0:
```diff
@@ -0,0 +1 @@
+?�7?�N?w}?H��>QY%?n�><Kd?��v?rR�>~�J?�e?^k?��l?Z{�=�p�=
��<�&U?H5G?�^?��z?��L?G�>��G?�9�=
```
Please mark these files as binary to avoid non-utf8 text warnings. 

https://stackoverflow.com/questions/7994811/git-mark-file-as-binary-to-avoid-line-separator-conversion

mciprian13 replied:
Is this ok (all the files in "tfliteModels" directory are binaries including the models themselves)?
```
* binary
```

yinghai replied:
The goal is when you update the PR, all the inp0 and model files should show up as binary. For example, in current status some are:
tests/models/tfliteModels/greater.inp0

But tests/models/tfliteModels/depthwise_conv2d_c1_m2.inp0 is not.

mciprian13 replied:
And where do I look to see whether the files are marked as binary or not?

mciprian13 replied:
I have no clue how to do it.

jackm321 replied:
@mciprian13, I need to confirm this is the issue on these files still, but I think the issue here is that both git and our internal linters use presence of NUL to determine if a file is binary or not and for the files that aren't marked as binary probably don't contain a NUL. Changing the git config could make git realize these are binary but won't help with our linters. Is there alternative way we can encode or otherwise deal with these files to fix this?

mciprian13 replied:
@jackm321 @yinghai I modified the file dumps to include a leading NULL byte (which of course is skipped when loading the data in the importer tests). Right now GIT recognizes all the files as being binary. Hopefully same thing happens in your internal linters.
Let's have this imported and hopefully landed (the OpenCL seems to fail for some reason not related to this PR).
Thanks!

jackm321 replied:
@mciprian13 that has fixed it, thanks for fixing that!
